File: Remote/Source/Networking/ISYConnectionManager.m
//
//  ISYConnectionManager.m
//  Remote
//
//  Created by Jason Cardwell on 9/3/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

#import "ISYConnectionManager.h"
#import "NetworkDeviceMulticastConnection.h"
#import "ISYDevice.h"
#import "Command.h"
#import "ISYDeviceConnection.h"
#import "ConnectionManager.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = (LOG_CONTEXT_NETWORKING | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel, msLogContext)

MSSTATIC_STRING_CONST kISYMulticastGroupAddress = @"239.255.255.250";
MSSTATIC_STRING_CONST kISYMulticastGroupPort    = @"1900";

@interface ISYConnectionManager () <NetworkDeviceConnectionDelegate>

@property (strong) NSMutableDictionary              * requestLog;           // Holds completion handlers
@property (strong) NSMutableArray                   * networkDevices;       // previously discovered devices.
@property (strong) NSMutableArray                   * deviceConnections;    // currently connected devices.
@property (strong) NSMutableSet                     * beaconsReceived;      // uuids  from processed beacons.
@property (copy)   NSString                         * capturedCommand;      // needs to be moved to `IRLearner`
@property (strong) NetworkDeviceMulticastConnection * multicastConnection;  // multicast group connection
@property (strong) NSHashTable                      * suspendedConnections; // active connections suspended

@property (assign) BOOL multicastGroupActive;                               // tracks whether should be in group

@property (strong) NSMutableDictionary * connectionCallbacks;               // device connection callbacks

@end

@implementation ISYConnectionManager

/**

 Accessor for the singleton instance of `GlobalCacheConnectionManager`

 @return const GlobalCacheConnectionManager *

 */
+ (const ISYConnectionManager *)sharedManager {

  static dispatch_once_t pred = 0;
  static const ISYConnectionManager * manager = nil;

  dispatch_once(&pred,  ^{

    /// Intialize shared manager and the manager's collection ivars
    ////////////////////////////////////////////////////////////////////////////////

    manager = [self new];

    manager.requestLog           = [@{} mutableCopy];
    manager.deviceConnections    = [@[] mutableCopy];
    manager.connectionCallbacks  = [@{} mutableCopy];
    manager.networkDevices       = [[ISYDevice findAll] mutableCopy] ?: [@[] mutableCopy];
    manager.beaconsReceived      = [NSMutableSet setWithCapacity:5];
    manager.suspendedConnections = [NSHashTable weakObjectsHashTable];

    /// Initialize the manager's multicast connection with a message handler
    ////////////////////////////////////////////////////////////////////////////////

    manager.multicastConnection =
    [NetworkDeviceMulticastConnection connectionWithAddress:kISYMulticastGroupAddress
                                                       port:kISYMulticastGroupPort
                                                   delegate:manager];


  });

  return manager;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Device Discovery
////////////////////////////////////////////////////////////////////////////////


/**

 Join multicast group and listen for beacons broadcast by iTach devices.

 @param completion Block to be executed upon completion of the task.

 */
+ (void)startDetectingDevices:(void(^)(BOOL success, NSError *error))completion {

  MSLogInfoTag(@"");
  
  // Set group active flag
  [self sharedManager].multicastGroupActive = YES;

  // Check for wifi
  if (![ConnectionManager isWifiAvailable]) {

    MSLogErrorTag(@"cannot detect network devices without valid wifi connection");

    if (completion) completion(NO, [NSError errorWithDomain:ConnectionManagerErrorDomain
                                                       code:ConnectionManagerErrorNoWifi
                                                   userInfo:nil]);

  }

  // Just execute the completion block if we have already joined
  else if ([self sharedManager].multicastConnection.isConnected) {

    MSLogWarnTag(@"multicast socket already exists");

    if (completion) completion(YES, nil);

  }

  // Otherwise join the multicast group
  else [[self sharedManager].multicastConnection connect:completion];

}

/**

 Cease listening for beacon broadcasts and release resources.

 @param completion Block to be executed upon completion of the task.

 */
+ (void)stopDetectingDevices:(void(^)(BOOL success, NSError *error))completion {

  MSLogInfoTag(@"");

  // Set group active flag
  [self sharedManager].multicastGroupActive = NO;

  // Leave group if joined to one
  if ([self sharedManager].multicastConnection.isConnected)
    [[self sharedManager].multicastConnection disconnect:completion];

  // Otherwise just execute completion block
  else {

    MSLogWarnTag(@"not currently joined to a multicast group");
    if (completion) completion(YES, nil);

  }

}

/**
 Whether socket is open to receive multicast group broadcast messages.

 @return `YES` if detecting and `NO` otherwise.

 */
+ (BOOL)isDetectingNetworkDevices {
  //???: Should this return whether flag is set for multicast group or whether connection exists?
  return [self sharedManager].multicastGroupActive;
}


////////////////////////////////////////////////////////////////////////////////
#pragma mark - Connecting/sending to a device
////////////////////////////////////////////////////////////////////////////////


/**

 Sends an IR command to the device identified by the specified `uuid`.

 @param command The command to execute

 @param completion The block to execute upon task completion

 */
+ (void)sendCommand:(HTTPCommand *)command completion:(void (^)(BOOL success, NSError *))completion {


}


/// Suspend active connections
+ (void)suspend {

  if ([self sharedManager].multicastGroupActive)
    [[self sharedManager].multicastConnection disconnect:nil];

}

/// Resume previously active connections
+ (void)resume {

  if ([self sharedManager].multicastGroupActive)
    [[self sharedManager].multicastConnection disconnect:nil];

}



////////////////////////////////////////////////////////////////////////////////
#pragma mark - NetworkDeviceConnectionDelegate
////////////////////////////////////////////////////////////////////////////////

- (void)deviceDisconnected:(NetworkDeviceConnection *)connection {
  MSLogInfoTag(@"");
}
- (void)deviceConnected:(NetworkDeviceConnection *)connection {
  MSLogInfoTag(@"");
}

- (void)messageReceived:(NSString *)message overConnection:(NetworkDeviceConnection *)connection {

  MSLogInfoTag(@"message received:\n%@\n", message);

  MSDictionary * entries =
  [MSDictionary dictionaryByParsingArray:[message matchingSubstringsForRegEx:@"^[A-Z]+:.*(?=\\r)"]];
  MSLogInfo(@"entries: %@", entries);

  NSString * location = entries[@"LOCATION"];

  // Check if device has already been discovered
  if ([location hasSuffix:@"/desc"] && ![self.beaconsReceived containsObject:location]) {

    // Add to list of received beacons
    [self.beaconsReceived addObject:location];

    NSString * baseURL = [location stringByReplacingOccurrencesOfString:@"/desc" withString:@""];
    __weak ISYConnectionManager * weakself = self;

    [ISYDeviceConnection connectionWithBaseURL:[NSURL URLWithString:baseURL]
                                    completion:^(ISYDeviceConnection *connection)
     {
       assert(IsMainQueue);

       if (connection) {

         assert([weakself.deviceConnections firstObjectCommonWithArray:@[connection]] == nil);
         [weakself.deviceConnections addObject:connection];
         [[weakself class] stopDetectingDevices:nil];

       }

     }];

  }

}

- (void)messageSent:(NSString *)message overConnection:(NetworkDeviceConnection *)connection {
  MSLogInfoTag(@"message: %@", message);
}


@end

File: Networking/Source/ISYConnectionManager.swift
//
//  ISYConnectionManager.swift
//  Remote
//
//  Created by Jason Cardwell on 5/06/15.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit
import Settings
import class DataModel.HTTPCommand
import class DataModel.ISYDevice
import class DataModel.DataManager
import class DataModel.NetworkDevice

final class ISYConnectionManager {

  typealias Callback = ConnectionManager.Callback
  typealias Error = ConnectionManager.Error
  typealias Device = ISYDevice
  typealias DeviceIdentifier = String
  typealias Connection = ISYDeviceConnection

  static let MulticastAddress = "239.255.255.250"
  static let MulticastPort: UInt16 = 1900
  static let MulticastSearchMessage = "M-SEARCH * HTTP/1.1\rHOST:239.255.255.250:1900\rMAN:\"ssdp.discover\"\rMX:1\rST:urn:udi-com:device:X_Insteon_Lighting_Device:1"

  static private var context: NSManagedObjectContext = DataManager.rootContext

  /** Previously discovered devices. */
  static private(set) var networkDevices = Set(ISYDevice.objectsInContext(DataManager.rootContext) as! [ISYDevice])

  /** Current/previous device connections */
  static private var connections: [DeviceIdentifier:Connection] = [:]

  /** Uuids from processed beacons. */
  static private var beaconsReceived: Set<String> = []

  /** Multicast group connection */
  static let multicastConnection = MulticastConnection(address:ISYConnectionManager.MulticastAddress,
                                                       port: ISYConnectionManager.MulticastPort,
                                                       didReceiveMessage: ISYConnectionManager.messageReceived)

  /** Whether socket is open to receive multicast group broadcast messages */
  static private(set) var detectingNetworkDevices = false

  /**
  Join multicast group and listen for beacons broadcast by iTach devices.

  - parameter context: NSManagedObjectContext = DataManager.rootContext
  */
  class func startDetectingNetworkDevices(context: NSManagedObjectContext = DataManager.rootContext) throws {
    guard !detectingNetworkDevices else { return }
    self.context = context
    detectingNetworkDevices = true
    try multicastConnection.listen()
    multicastConnection.sendMessage(MulticastSearchMessage)
    MSLogDebug("listening for ISY devices…")
  }

  /** Cease listening for beacon broadcasts and release resources. */
  class func stopDetectingNetworkDevices() {
    guard detectingNetworkDevices else { return }
    multicastConnection.stopListening()
    detectingNetworkDevices = false
    MSLogDebug("no longer listening for ISY devices")
  }

  /** Suspend active connections */
  class func suspend() { guard detectingNetworkDevices else { return }; multicastConnection.stopListening() }

  /** Resume previously active connections */
  class func resume() throws { guard detectingNetworkDevices else { return };  try multicastConnection.listen() }

  /**
  Processes messages received through `NetworkDeviceConnection` objects.

  - parameter message: String Contents of the message received by the device connection
  */
  class func messageReceived(message: String) {

    MSLogDebug("message received over multicast connection:\n\(message)\n")

    guard let uniqueIdentifier = (~/"USN:(uuid:[0-9a-f:]{17})").firstMatch(message)?.captures[1]?.string
      where beaconsReceived ∌ uniqueIdentifier else { return }

    beaconsReceived.insert(uniqueIdentifier)

    guard let location = (~/"LOCATION:(http://[0-9.]+/desc)").firstMatch(message)?.captures[1]?.string,
      descURL = NSURL(string: location) else { return }

    NSURLSession(configuration: NSURLSessionConfiguration.ephemeralSessionConfiguration()).dataTaskWithURL(descURL) {
      (data: NSData?, response: NSURLResponse?, error: NSError?) -> Void in

      guard let data = data else { MSHandleError(error); return }
      
      let isUpdate = Device.objectExistsInContext(context, withValue: uniqueIdentifier, forAttribute: "uniqueIdentifier")

      guard let device = Device.deviceFromDesc(data, context: context) else { return }

      let notify = isUpdate ? ConnectionManager.updatedDevice : ConnectionManager.discoveredDevice
      let shouldStopKey: ConnectionManager.SettingKey = isUpdate ? .StopAfterUpdated : .StopAfterDiscovered
      let shouldConnectKey: ConnectionManager.SettingKey = isUpdate ? .AutoConnectExisting : .AutoConnectDiscovery

      notify(device)

      // Stop if appropriate
      if SettingsManager.valueForSetting(shouldStopKey) == true { stopDetectingNetworkDevices() }

      // Connect if appropriate
      if SettingsManager.valueForSetting(shouldConnectKey) == true { connections[uniqueIdentifier] = Connection(device: device) }
    } .resume()

  }

}